bebraven/platform
# Handle ID attributes for headings (and sometimes other things too)

### Description

Add general element ID handling, and custom heading-specific ID handling.

This does not check uniqueness, it just generates UUIDs and relies on them being unique.

Task: https://app.asana.com/0/1174274412967132/1195096839523739/f

### Testing

1. Open the editor with the inspector (/custom_contents/new?debug)
1. Type something in a document
2. From the block toolbar (left **P** button), change the Paragraph dropdown to any Heading.
3. See that there's an ID attribute on the newly inserted heading.
4. Try to break things (toggle code tab, back to design tab, press enter after the heading, change the heading type from the dropdown, ???), notice the IDs remain unchanged
5. Change the heading to a paragraph, notice the ID is removed
5. Copy/paste the heading (might have to mess with it so that the whole heading element is inside the selection), notice the ID changes
6. Click on the heading, notice the right panel with copyable (but read-only) heading anchor link.


### Known issues

* Press Enter before the heading (heading with duplicate ID created)
* Press Enter in the middle of the heading, splitting it in two (heading with duplicate ID created)



### Motivation and Context




### Screenshots (if appropriate):

<img width="395" alt="Screen Shot 2020-11-11 at 3 19 28 PM" src="https://user-images.githubusercontent.com/1382374/98865631-4e43dd80-2431-11eb-859f-b79eea1924da.png">
<img width="1237" alt="Screen Shot 2020-11-11 at 3 25 47 PM" src="https://user-images.githubusercontent.com/1382374/98866209-3456ca80-2432-11eb-9ea1-27c70b505971.png">


### Types of changes




### Checklist:
